### DISCUSSION TITLE
Use new OpenAI feature for parsing ingredients or even recipes

### DISCUSSION THREAD

**Discussion by L3-N0X** [mealie-recipes/mealie#3999] [Feature Request]
### First Check

- [X] I used the GitHub search to find a similar requests and didn't find it.
- [X] Checked the [tasks tagged](https://github.com/mealie-recipes/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

### Please provide a concise description of the problem that would be addressed by this feature.

The parsing with the current implementation of OpenAI is not really useful and does not work all the time. Now OpenAI released a new tool called Structured Outputs in the API and I think this would be perfect for this application, even better than the current parser(s). This may also lead to make the workflow of auto-parsing possible (currently not available because parsers are unreliable)

I would really love to see this feature, because parsing each recipe by hand is really annoying and the OpenAI parser, which always uses like 7 requests, is not very economical and cheap and 1 single request would be much better. 

I would also like it when it would be possible to use any OpenAI-API compatible Endpoint, because this would enable using other services which provide the same endpoints, for example a ollama installation on my PC in the LAN, providing such an API on <ip>:11343/v1
But this is another topic.

Here is the post of the new OpenAI feature, published on 06.08.2024
https://openai.com/index/introducing-structured-outputs-in-the-api/

### Please provide a concise description of the feature that would resolve your issue.

Implement the new Structured Outputs feature as a parser in mealie

### Please consider and list out some caveats or tradeoffs made in your design decision

Only OpenAI is offering this feature right now and there may be a new endpoint/breaking api change only for structured outputs, but I don't think this is likely to happen

### Additional Information

- [ ] If this is accepted I'm willing to submit a PR to provide this feature
- [ ] If this is accepted I'm willing to help maintain this feature
- [ ] I'm willing to sponsor/pay a developer to do this work

**Comment by michael-genson** [comment]
> The parsing with the current implementation of OpenAI is not really useful and does not work all the time

Can you go into more detail? What issues are you having with it? Curious to see some examples where it's failing, we might need to tweak the prompt. I've been using it exclusively over the NLP parser and it's been working wonders for me.

The new structured outputs setting is only for forcing the model to output in an expected JSON format, which it already does. Results won't be any different than they are now; if anything it would just simplify the codebase.

> parsing each recipe by hand is really annoying and the OpenAI parser, which always uses like 7 requests

This is a configuration setting, see `OPENAI_WORKERS`: https://docs.mealie.io/documentation/getting-started/installation/backend-config/#openai

> I would also like it when it would be possible to use any OpenAI-API compatible Endpoint

This is also a configuration setting, see `OPENAI_BASE_URL` and `OPENAI_API_KEY`. Some users are using ollama, though I haven't heard how successful it's been.

> This may also lead to make the workflow of auto-parsing possible

Completely agree here. There is some additional work to be done, I think the ideal state builds the workflow into new recipe creation (since auto-parsing will _always_ have some errors). @hay-kot has an awesome workflow with Recipinned that I think we can at least partially adopt with Mealie.

**Reply by ghostwolf1228** [reply]
@michael-genson any update on the auto-parsing feature? Even if it has errors, it appears to be correct more frequently than it is wrong and fixing one or two incorrect entries is a heck of a lot easier than doing the whole thing for each recipe. This would also come in handy if we wanted to do a bulk import of recipes (especially in my case, I'm going around to various sites and importing stuff and I'd have a lot more in my instance if auto-parsing was a thing).

**Reply by L3-N0X** [reply]
Especially the new gpt4.1 Models with their big context window of 1 million tokens should work great for this, you can send all the previously created ingredients, send the whole web page/Image and the json schema and the AI would return everything parsed correctly. This should also work just fine on the smaller Models like 4.1 Mini which makes it cheaper and you could even use geminis free api quota for this too.
Would love to see this feature, I don't have the time to create an PR for this currently.

**Reply by michael-genson** [reply]
I started working on part of this implementation but anything that touches the frontend is on pause right now for [the Nuxt 3 upgrade](https://github.com/mealie-recipes/mealie/pull/5184)
